larq_compute_engine/mlir/transforms/optimize_patterns.td

```diff
@@ -41,9 +41,23 @@ multiclass FuseMulOrDivWithBConv2D<dag binaryOp> {
                                   (ConstantOp $value), TFL_AF_None),
                         $channels_in, $strides, $padding, $pad_values,
                         $dilations, $filter_format, $activation),
-         [(HasOneUse $conv_output)]>;
+         [(HasOneUse $conv_output)], (addBenefit 100)>;
 }
 
+def HasNegativeValues : AttrConstraint<CPred<"HasNegativeValues($_self)">>;
+def ComputeBSignAndExpandTo4D : NativeCodeCall<"ComputeBSignAndExpandTo4D($0)">;
+
+def : Pat<(TF_LceBconv2dOp $input, (ConstantOp F32ElementsAttr:$filter),
+            (ConstantOp HasNegativeValues:$post_activation_multiplier),
+            $post_activation_bias, $channels_in, $strides, $padding,
+            $pad_values, $dilations, $filter_format, $act),
+          (TF_LceBconv2dOp $input,
+            (TFL_MulOp (ConstantOp $filter), (ConstantOp (ComputeBSignAndExpandTo4D $post_activation_multiplier)), TFL_AF_None),
+            (TFL_AbsOp (ConstantOp $post_activation_multiplier)),
+            $post_activation_bias, $channels_in, $strides, $padding,
+            $pad_values, $dilations, $filter_format, $act),
+          [], (addBenefit 90)>;
```

Comment:
Really nice that you can just use the `MulOp` for this and don't have to manually iterate through the weights.
Some questions:
- The `addBenefit` numbers that you changed, are those to make sure that this happens after the batchnorm has been merged?
- How do you make sure this pattern happens before bitpacking the filters?
- I think this pattern should only be applied when there is no activation function, so the `$act` should be replaced by `TFL_AF_None`.

Reply:
> Really nice that you can just use the MulOp for this and don't have to manually iterate through the weights.

Yes, if we would implement constant folding for the bsignop, we could've also used it here as well.

> The `addBenefit` numbers that you changed, are those to make sure that this happens after the batchnorm has been merged?

Its main goal is to make sure that the transformation happens before the bitpacking the weights, since we can't apply the transformation after it. In most cases this will also make the pattern run after batchnorm fusing, but this wouldn't be strickly necessary since the pattern will be applied again if batchnorm fusing would reintroduce negative multipliers.

> How do you make sure this pattern happens before bitpacking the filters?

`F32ElementsAttr:$filter` makes sure it doesn't run afterwards and `addBenefit` makes sure that it runs before. I guess in the future it probably makes sense to extract the bitpacking related transforms into their own pass that runs afterwards which would also make testing much easier.

> I think this pattern should only be applied when there is no activation function, so the `$act` should be replaced by `TFL_AF_None`.

Good catch, thanks for noticing that! I fixed it in 1d4b29ba6cfd0d9865035fb398ca86d07664ade3
I was distracted by some bitpacking related bugs so this slipped by when pushing the PR.